Unit flake in contrib/mesos/pkg/offers

From http://pr-test.k8s.io/23649/kubernetes-pull-test-unit-integration/19981/

```
panic: test timed out after 5m0s

goroutine 3675 [running]:
panic(0x1299dc0, 0xc8201ec8e0)
	/usr/local/go/src/runtime/panic.go:464 +0x3ff
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:725 +0x184
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive, 1 minutes]:
testing.RunTests(0x18e4520, 0x1f4a000, 0x5, 0x5, 0xc8201ed001)
	/usr/local/go/src/testing/testing.go:583 +0xb39
testing.(*M).Run(0xc82006df00, 0x0)
	/usr/local/go/src/testing/testing.go:515 +0x11e
main.main()
	k8s.io/kubernetes/contrib/mesos/pkg/offers/_test/_testmain.go:62 +0x211

goroutine 17 [syscall, 1 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 19 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x1f69c80)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x8c
created by github.com/golang/glog.init.1
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2ba

goroutine 5 [semacquire, 1 minutes]:
sync.runtime_Syncsemacquire(0xc82024a040)
	/usr/local/go/src/runtime/sema.go:241 +0x201
sync.(*Cond).Wait(0xc82024a030)
	/usr/local/go/src/sync/cond.go:63 +0xb3
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).pop.func1.1(0xc82001a180, 0xc82023e0a0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:349 +0x7a
created by k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).pop.func1
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:350 +0x13e

goroutine 26 [chan receive, 1 minutes]:
testing.(*T).Parallel(0xc8200cf7a0)
	/usr/local/go/src/testing/testing.go:441 +0x1e4
k8s.io/kubernetes/contrib/mesos/pkg/offers.TestExpiredOffer(0xc8200cf7a0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers_test.go:31 +0x51
testing.tRunner(0xc8200cf7a0, 0x1f4a000)
	/usr/local/go/src/testing/testing.go:473 +0xdd
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0xae3

goroutine 27 [chan receive, 1 minutes]:
testing.RunTests.func1(0xc820093aa0, 0xc8200cf7a0)
	/usr/local/go/src/testing/testing.go:586 +0x6c
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:587 +0xbc5

goroutine 28 [chan receive, 1 minutes]:
testing.(*T).Parallel(0xc8200cf830)
	/usr/local/go/src/testing/testing.go:441 +0x1e4
k8s.io/kubernetes/contrib/mesos/pkg/offers.TestTimedOffer(0xc8200cf830)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers_test.go:57 +0x51
testing.tRunner(0xc8200cf830, 0x1f4a018)
	/usr/local/go/src/testing/testing.go:473 +0xdd
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0xae3

goroutine 29 [chan receive, 1 minutes]:
testing.RunTests.func1(0xc820093aa0, 0xc8200cf830)
	/usr/local/go/src/testing/testing.go:586 +0x6c
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:587 +0xbc5

goroutine 56 [chan receive, 1 minutes]:
k8s.io/kubernetes/contrib/mesos/pkg/offers.TestListen(0xc82027c3f0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers_test.go:284 +0x56c
testing.tRunner(0xc82027c3f0, 0x1f4a048)
	/usr/local/go/src/testing/testing.go:473 +0xdd
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:582 +0xae3

goroutine 39 [select, 1 minutes]:
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).pop.func1(0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:351 +0x259
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop(0xc82024a000, 0xc820061bb8, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:146 +0x75
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).pop(0xc82023e0a0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:376 +0xd7
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).Pop(0xc82023e0a0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:332 +0x49
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).nextListener(0xc8202382d0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:444 +0x6e
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).notifyListeners(0xc8202382d0, 0xc820061e90)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:457 +0x4f
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).Init.func2()
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:508 +0x84
k8s.io/kubernetes/contrib/mesos/pkg/runtime.Until.func1(0xc82023e0c0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/runtime/util.go:115 +0x59
k8s.io/kubernetes/contrib/mesos/pkg/runtime.Until(0xc82023e0c0, 0x0, 0xc820236180)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/runtime/util.go:116 +0x9a
created by k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).Init
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:508 +0x298

goroutine 31 [semacquire, 1 minutes]:
sync.runtime_Syncsemacquire(0xc8202820b0)
	/usr/local/go/src/runtime/sema.go:241 +0x201
sync.(*Cond).Wait(0xc8202820a0)
	/usr/local/go/src/sync/cond.go:63 +0xb3
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop.func1(0xc820282070, 0xc8200a01f8)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:161 +0x96
created by k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:163 +0x20d

goroutine 10 [semacquire, 1 minutes]:
sync.runtime_Syncsemacquire(0xc82024a0b0)
	/usr/local/go/src/runtime/sema.go:241 +0x201
sync.(*Cond).Wait(0xc82024a0a0)
	/usr/local/go/src/sync/cond.go:63 +0xb3
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop.func1(0xc82024a070, 0xc820050038)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:161 +0x96
created by k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:163 +0x20d

goroutine 57 [semacquire, 1 minutes]:
sync.runtime_Syncsemacquire(0xc8202820b0)
	/usr/local/go/src/runtime/sema.go:241 +0x201
sync.(*Cond).Wait(0xc8202820a0)
	/usr/local/go/src/sync/cond.go:63 +0xb3
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).Pop.func1(0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:133 +0xdf
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop(0xc820282070, 0xc820275df8, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:146 +0x75
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).Pop(0xc820282070, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:138 +0x86
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).ageOffers(0xc82027a320)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:429 +0x56
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).(k8s.io/kubernetes/contrib/mesos/pkg/offers.ageOffers)-fm()
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:492 +0x2e
k8s.io/kubernetes/contrib/mesos/pkg/runtime.Until.func1(0xc82025e260)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/runtime/util.go:115 +0x59
k8s.io/kubernetes/contrib/mesos/pkg/runtime.Until(0xc82025e260, 0x0, 0xc8202604e0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/runtime/util.go:116 +0x9a
created by k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).Init
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:492 +0xbf

goroutine 58 [select]:
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop(0xc820282000, 0xc820273bb8, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:164 +0x3f0
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).pop(0xc8202622c0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:376 +0xd7
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayFIFO).Pop(0xc8202622c0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:332 +0x49
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).nextListener(0xc82027a320, 0x2)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:444 +0x6e
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).notifyListeners(0xc82027a320, 0xc820273e90)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:457 +0x4f
k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).Init.func2()
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:508 +0x84
k8s.io/kubernetes/contrib/mesos/pkg/runtime.Until.func1(0xc8202622e0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/runtime/util.go:115 +0x59
k8s.io/kubernetes/contrib/mesos/pkg/runtime.Until(0xc8202622e0, 0x0, 0xc8202604e0)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/runtime/util.go:116 +0x9a
created by k8s.io/kubernetes/contrib/mesos/pkg/offers.(*offerStorage).Init
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/offers/offers.go:508 +0x298

goroutine 3682 [semacquire]:
sync.runtime_Syncsemacquire(0xc820282040)
	/usr/local/go/src/runtime/sema.go:241 +0x201
sync.(*Cond).Wait(0xc820282030)
	/usr/local/go/src/sync/cond.go:63 +0xb3
k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop.func1(0xc820282000, 0xc820240738)
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:161 +0x96
created by k8s.io/kubernetes/contrib/mesos/pkg/queue.(*DelayQueue).pop
	/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/contrib/mesos/pkg/queue/delay.go:163 +0x20d
FAIL	k8s.io/kubernetes/contrib/mesos/pkg/offers	300.337s
```

--- a/contrib/mesos/pkg/queue/delay.go
+++ b/contrib/mesos/pkg/queue/delay.go
@@ -18,6 +18,7 @@ package queue
 
 import (
 	"container/heap"
+	"runtime"
 	"sync"
 	"time"
 
@@ -138,6 +139,29 @@ func (q *DelayQueue) Pop() interface{} {
 	}, nil)
 }
 
+func finishWaiting(cond *sync.Cond, waitFinished <-chan struct{}) {
+	runtime.Gosched()
+	select {
+	// avoid creating a timer if we can help it...
+	case <-waitFinished:
+		return
+	default:
+		const spinTimeout = 100 * time.Millisecond
+		t := time.NewTimer(spinTimeout)
+		defer t.Stop()
+		for {
+			runtime.Gosched()
+			cond.Broadcast()
+			select {
+			case <-waitFinished:
+				return
+			case <-t.C:
+				t.Reset(spinTimeout)
+			}
+		}
+	}
+}
+
 // returns a non-nil value from the queue, or else nil if/when cancelled; if cancel
 // is nil then cancellation is disabled and this func must return a non-nil value.
 func (q *DelayQueue) pop(next func() *qitem, cancel <-chan struct{}) interface{} {
@@ -164,6 +188,7 @@ func (q *DelayQueue) pop(next func() *qitem, cancel <-chan struct{}) interface{}
 			select {
 			case <-cancel:
 				item.readd(item)
+				finishWaiting(&q.cond, ch)
 				return nil
 			case <-ch:
 				// we may no longer have the earliest deadline, re-try
@@ -353,8 +378,7 @@ func (q *DelayFIFO) pop(cancel <-chan struct{}) interface{} {
 					// we may not have the lock yet, so
 					// broadcast to abort Wait, then
 					// return after lock re-acquisition
-					q.cond().Broadcast()
-					<-signal
+					finishWaiting(q.cond(), signal)
 					return nil
 				case <-signal:
 					// we have the lock, re-check

